Incompatible with Go 1.23 due to locking down `linkname` using

**Describe the bug**

Upstream JSON lib `sonic` is suffering from `linkname` usage locking down. And it affects hertz according to reports.

              got the same issue from `cloudwego/hertz`.

_Originally posted by @xin-hao-awx in https://github.com/bytedance/sonic/issues/660#issuecomment-2190483847_
            

Also can find more information at https://github.com/bytedance/sonic/issues/637

BTW, `linkname` is used in hertz, too. So maybe we need an investigation.

**To Reproduce**

1. Download Go 1.23 RC1
2. Build a naive Hertz project
3. See errors in building phase

**Expected behavior**

No error should be presented.

**Hertz version:**

On `develop` branch

**Environment:**

```plain
GO111MODULE=''
GOARCH='amd64'
GOBIN=''
GOCACHE='/home/test/.cache/go-build'
GOENV='/home/test/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/home/test/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/home/test/go'
GOPRIVATE=''
GOPROXY='https://goproxy.cn,direct'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/usr/local/go/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.23rc1'
GODEBUG=''
GOTELEMETRY='local'
GOTELEMETRYDIR='/home/test/.config/go/telemetry'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='gcc'
CXX='g++'
CGO_ENABLED='0'
GOMOD='/home/test/xcpc-team-reg-main/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -m64 -fno-caret-diagnostics -Qunused-arguments -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build2791510403=/tmp/go-build -gno-record-gcc-switches'
```

**Additional context**

We have an update PR (#1106 ) but idle for a time. I can compile a naive hertz project with Go 1.21, and it runs OK. Please consider bumping the PR status.

> refer to https://www.cloudwego.io/zh/docs/hertz/tutorials/basic-feature/json/
> 

For the following code:

```go
package main

import (
	"context"

	"github.com/cloudwego/hertz/pkg/app"
	"github.com/cloudwego/hertz/pkg/app/server"
	"github.com/cloudwego/hertz/pkg/common/utils"
	"github.com/cloudwego/hertz/pkg/protocol/consts"
)

func main() {
	h := server.Default(server.WithHostPorts("12345"))

	h.GET("/ping", func(c context.Context, ctx *app.RequestContext) {
		ctx.JSON(consts.StatusOK, utils.H{"ping": "pong"})
	})

	h.Spin()
}
```

Compiling with Go 1.23-rc1:

```bash
go build -tags stdjson
```

and the error message is:

```plain
# github.com/bytedance/sonic/loader
../go/pkg/mod/github.com/bytedance/sonic@v1.11.3/loader/funcdata.go:37:32: undefined: _func
../go/pkg/mod/github.com/bytedance/sonic@v1.11.3/loader/funcdata.go:74:12: undefined: moduledata
../go/pkg/mod/github.com/bytedance/sonic@v1.11.3/loader/funcdata.go:77:18: undefined: moduledata
../go/pkg/mod/github.com/bytedance/sonic@v1.11.3/loader/funcdata.go:188:28: undefined: compilationUnit
../go/pkg/mod/github.com/bytedance/sonic@v1.11.3/loader/stubs.go:27:22: undefined: moduledata
../go/pkg/mod/github.com/bytedance/sonic@v1.11.3/loader/stubs.go:29:26: undefined: moduledata
../go/pkg/mod/github.com/bytedance/sonic@v1.11.3/loader/stubs.go:34:27: undefined: moduledata
../go/pkg/mod/github.com/bytedance/sonic@v1.11.3/loader/stubs.go:36:36: undefined: moduledata
../go/pkg/mod/github.com/bytedance/sonic@v1.11.3/loader/stubs.go:46:21: undefined: moduledata
../go/pkg/mod/github.com/bytedance/sonic@v1.11.3/loader/stubs.go:50:22: undefined: moduledata
../go/pkg/mod/github.com/bytedance/sonic@v1.11.3/loader/stubs.go:50:22: too many errors
```

--- a/go.mod
+++ b/go.mod
@@ -6,7 +6,7 @@ require (
 	github.com/bytedance/go-tagexpr/v2 v2.9.2
 	github.com/bytedance/gopkg v0.1.0
 	github.com/bytedance/mockey v1.2.10
-	github.com/bytedance/sonic v1.11.9
+	github.com/bytedance/sonic v1.12.0
 	github.com/cloudwego/netpoll v0.6.2
 	github.com/fsnotify/fsnotify v1.5.4
 	github.com/tidwall/gjson v1.14.4
@@ -16,7 +16,7 @@ require (
 )
 
 require (
-	github.com/bytedance/sonic/loader v0.1.1 // indirect
+	github.com/bytedance/sonic/loader v0.2.0 // indirect
 	github.com/cloudwego/base64x v0.1.4 // indirect
 	github.com/cloudwego/iasm v0.2.0 // indirect
 	github.com/golang/protobuf v1.5.0 // indirect

